### PR TITLE
feat(pinecone): pinecone batch upsert

### PR DESCRIPTION
**Pull request by chuang8511**
Because

- we will need batch upsert

This commit

- refactor task upsert with concurrent execution

### CODE REVIEW

File: pkg/component/data/pinecone/v0/component_test.go [hunk 1]
```diff
@@ -253,64 +229,4 @@ func TestComponent_Execute(t *testing.T) {
 		})
 	}
 
-	c.Run("nok - 400", func(c *qt.C) {
```
Comment (jvallesm) [inline pkg/component/data/pinecone/v0/component_test.go:256]: These error tests should work with other tasks, too. We might not need to remove them.

File: pkg/component/data/pinecone/v0/component_test.go [hunk 2]
```diff
@@ -105,20 +95,6 @@ func TestComponent_Execute(t *testing.T) {
 		wantClientReq  any
 		clientResp     string
 	}{
```
Comment (jvallesm) [inline pkg/component/data/pinecone/v0/component_test.go:97]: 🙏 Can you add a TODO to reintroduce this testcase?

```go
// TODO: #927 removed the `taskUpsert` testcase. Reintroduce it in INS-7102.

File: pkg/component/data/pinecone/v0/io.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+	Metadata  interface{} `instill:"metadata"`
+	Values    []float64   `instill:"values"`
+	Namespace string      `instill:"namespace"`
+}
+
+type taskUpsertOutput struct {
+	UpsertedCount int64 `instill:"upserted-count"`
+}
+
+type taskBatchUpsertInput struct {
```
[the remark this answers is not in the document]
Comment (chuang8511) [reply pkg/component/data/pinecone/v0/io.go:15]: > which imply that they're ordered and there's a correspondence in their indices

Previously, in other component, like document component, we follow this strategy. So, I developed it by following previous design.

But, actually, I am not sure if we should follow this strategy for the future development.
I haven't thought deeply which way makes a better DX.

1. make input slices of data, which imply that they're ordered and there's a correspondence in their indices. 
e.g. 
input: []ids, []vectors, 
data: 
- ids[0] -> vectors[0]
- ids[1] -> vectors[1]
- ...

2. make input slices of object data.
input: []vectorsData
vectorsData stands for
- id
- vectors

cc @GeorgeWilliamStrong

File: pkg/component/data/pinecone/v0/component_test.go (hunk 2 above)
Comment (jvallesm) [reply pkg/component/data/pinecone/v0/component_test.go:108]: 🙇

File: pkg/component/data/pinecone/v0/io.go [hunk 4]
```diff
@@ -22,3 +22,9 @@ type taskBatchUpsertInput struct {
 type taskBatchUpsertOutput struct {
 	UpsertedCount int64 `instill:"upserted-count"`
 }
+
+type vector struct {
```
[the remark this answers is not in the document]
Comment (chuang8511) [reply pkg/component/data/pinecone/v0/io.go:24]: Thanks! I was thinking about the batch issue but forgot to update it.

File: pkg/component/data/pinecone/v0/io.go [hunk 5]
```diff
@@ -1,28 +1,23 @@
-// TASK_QUERY and TASK_RERANK are not refactored yet, they will be addressed in ins-7102
+// TODO: TASK_QUERY and TASK_RERANK are not refactored yet, they will be
+// addressed in INS-7102.
 package pinecone
 
+type vector struct {
+	ID       string            `json:"id" instill:"id"`
+	Values   []float64         `json:"values,omitempty" instill:"values"`
+	Metadata map[string]string `json:"metadata,omitempty" instill:"metadata"`
```
Comment (jvallesm) [inline pkg/component/data/pinecone/v0/io.go:8]: TODO: we need to make this `map[string]any` but `ReadData` fails (cc @donch1989)

File: pkg/component/data/pinecone/v0/io.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jvallesm) [reply pkg/component/data/pinecone/v0/io.go:16]: Added in 12d36fd5c44f75562e804e0170a7c74e42cb7d5c, plus the batch upsert tests

**Comment by jvallesm** [comment]
# QA

### Recipe

```yaml
variable:
  id:
    format: string
  colour:
    format: string
component:
  pinecone-0:
    type: pinecone
    task: TASK_BATCH_UPSERT
    input:
      vectors:
        - id: d
          values: [1.0, 2.0, 2.0]
          metadata:
            color: dusk
        - id: ${variable.id}
          values: [0.0, -1.0, 5.0]
          metadata:
            color: ${variable.colour}
      namespace: jazz
    condition:
    setup: ${connection.pinecone}
output:
  out:
    value: ${pinecone-0.output.upserted-count}
```

### Request

```sh
$ curl -X POST 'http://localhost:8080/v1beta/users/jvallesm/pipelines/pineco/trigger' \
--header "Content-Type: application/json" \
--header "Authorization: Bearer $INSTILL_API_TOKEN" \
--data '{
  "inputs": [
    {
      "colour": "orchid",
      "id": "z"
    }
  ]
}' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   247  100   172  100    75     45     19  0:00:03  0:00:03 --:--:--    64
{
  "outputs": [
    {
      "out": 2
    }
  ],
  "metadata": {
    "traces": {
      "pinecone-0": {
        "statuses": [
          "STATUS_COMPLETED"
        ],
        "inputs": [],
        "outputs": [],
        "error": {
          "message": ""
        },
        "computeTimeInSeconds": 0
      }
    }
  }
}
```

### Result
CleanShot 2024-12-10 at 19 58 36

**Pull request merged**